File: javascript/data-structures/p157_maxHeap.js
/*
    CLRS Section 6.3, p. 157
*/

class MaxHeap {
    constructor(A = [], heapSize = 0) {
        this.A = A;
        this.heapSize = heapSize;
    }

    parent = (i) => Math.floor(parseInt((i - 2) / 2));

    left = (i) => 2 * i + 1;

    right = (i) => 2 * i + 2;

    maxHeapify = (i) => {
        let l = this.left(i);
        let r = this.right(i);
        let largest;
        if ((l < this.heapSize) && (this.A[l] > this.A[i])) {
            largest = l;
        } else {
            largest = i;
        }
        if ((r < this.heapSize) && (this.A[r] > this.A[largest])) {
            largest = r;
        }
        if (largest !== i) {
            [this.A[i], this.A[largest]] = [this.A[largest], this.A[i]];
            this.maxHeapify(largest);
        }
    }

    buildMaxHeap = () => {
        this.heapSize = this.A.length;
        for(let i = Math.floor((this.A.length - 1)/2); i >= 0; i--) {
            this.maxHeapify(i);
        }
    }
}

module.exports = {
    MaxHeap
}

File: python/data_structures/p157_max_heap.py
'''
    CLRS Section 6.3, p. 157
'''

class MaxHeap:
    def __init__(self, A=None, heap_size = 0):
        if A is None:
            self.A = []
        else:
            self.A = A
        self.heap_size = heap_size
    
    def parent(self, i):
        return (i - 1) // 2
    
    def left(self, i):
        return 2 * i + 1
    
    def right(self, i):
        return 2 * i + 2
    
    def max_heapify(self, i):
        l = self.left(i)
        r = self.right(i)
        largest = None
        if l < self.heap_size and self.A[l] > self.A[i]:
            largest = l
        else:
            largest = i
        if r < self.heap_size and self.A[r] > self.A[largest]:
            largest = r
        if largest != i:
            self.A[i], self.A[largest] = self.A[largest], self.A[i]
            self.max_heapify(largest)

    def build_max_heap(self):
        self.heap_size = len(self.A)
        for i in range((len(self.A) - 1)//2, -1, -1):
            self.max_heapify(i)
